### PR TITLE
Add JavaTimeModule to ObjectMappers to support Java8 date/time serial…

### PR DESCRIPTION
**Pull request by uschi2000**
…ization

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/http-remoting/329)

### CODE REVIEW

File: extras/jackson-support/src/test/java/com/palantir/remoting1/ext/jackson/ObjectMappersTest.java [hunk 1]
```diff
@@ -68,4 +71,18 @@ public void serializeJdk8ModuleEmptyOptional() throws JsonProcessingException {
...
     }
+
+    @Test
+    public void testJdk8DateTimeSerialization() throws IOException {
+        assertThat(MAPPER.writeValueAsString(Duration.ofDays(2))).isEqualTo("172800.000000000"); // 2*24*60*60
+        assertThat(MAPPER.readValue("172800.000000000", Duration.class)).isEqualTo(Duration.ofDays(2));
+        assertThat(MAPPER.readValue("172800", Duration.class)).isEqualTo(Duration.ofDays(2));
+
+        ZonedDateTime now = ZonedDateTime.now().withZoneSameInstant(ZoneId.of("GMT"));
+        assertThat(MAPPER.readValue(MAPPER.writeValueAsString(now), ZonedDateTime.class)).isEqualTo(now);
```
Comment (markelliot) [inline extras/jackson-support/src/test/java/com/palantir/remoting1/ext/jackson/ObjectMappersTest.java]: Would also like to assert this becomes an ISO-8601 compatible string.

**Comment by uschi2000** [comment]
It does not. It becomes a double-ish thing. Hence the loss in precision.
It's arguably not really the behavior we want

**Comment by markelliot** [comment]
There's a config option which should enable it to be output in the right format. Maybe I misread the docs but I thought it was default (apparently not). 

As an alternative we can just roll our own thing.

**Comment by uschi2000** [comment]
Well, that was painful. The default settings in Jackson are pretty weird, need to disable to "features".

File: extras/jackson-support/src/test/java/com/palantir/remoting1/ext/jackson/ObjectMappersTest.java [hunk 2]
```diff
@@ -68,4 +73,24 @@ public void serializeJdk8ModuleEmptyOptional() throws JsonProcessingException {
...
+    @Test
+    public void testJdk8DateTimeSerialization() throws IOException {
+        Duration duration = Duration.ofMinutes(60 * 50 + 1); // 50h, 1min
+        ZonedDateTime zoneDateTime = ZonedDateTime.of(2001, 2, 3, 4, 5, 6, 7, ZoneId.of(ZoneId.SHORT_IDS.get("EST")));
+        OffsetDateTime offsetDateTime = OffsetDateTime.of(2001, 2, 3, 4, 5, 6, 7, ZoneOffset.ofHours(-5));
+
+        assertThat(MAPPER.writeValueAsString(duration)).isEqualTo("\"PT50H1M\"");
+        serDe(duration, Duration.class);
+
+        assertThat(MAPPER.writeValueAsString(offsetDateTime)).isEqualTo("\"2001-02-03T04:05:06.000000007-05:00\"");
```
Comment (markelliot) [inline extras/jackson-support/src/test/java/com/palantir/remoting1/ext/jackson/ObjectMappersTest.java]: Any way to drop the digits that don't affect precision?

**Review by markelliot — APPROVED** [review]

File: extras/jackson-support/src/test/java/com/palantir/remoting1/ext/jackson/ObjectMappersTest.java [hunk 3]
```diff
@@ -68,4 +73,24 @@ public void serializeJdk8ModuleEmptyOptional() throws JsonProcessingException {
...
     }
+
+    @Test
+    public void testJdk8DateTimeSerialization() throws IOException {
+        Duration duration = Duration.ofMinutes(60 * 50 + 1); // 50h, 1min
+        ZonedDateTime zoneDateTime = ZonedDateTime.of(2001, 2, 3, 4, 5, 6, 7, ZoneId.of(ZoneId.SHORT_IDS.get("EST")));
+        OffsetDateTime offsetDateTime = OffsetDateTime.of(2001, 2, 3, 4, 5, 6, 7, ZoneOffset.ofHours(-5));
+
+        assertThat(MAPPER.writeValueAsString(duration)).isEqualTo("\"PT50H1M\"");
+        serDe(duration, Duration.class);
```
Comment (schlosna) [inline extras/jackson-support/src/test/java/com/palantir/remoting1/ext/jackson/ObjectMappersTest.java]: do you want to test return value?

File: extras/jackson-support/src/main/java/com/palantir/remoting1/ext/jackson/ObjectMappers.java [hunk 4]
```diff
@@ -34,6 +37,9 @@ public static ObjectMapper guavaJdk7Jdk8() {
                 .registerModule(new GuavaModule())
                 .registerModule(new ShimJdk7Module())
                 .registerModule(new Jdk8Module())
-                .registerModule(new AfterburnerModule());
+                .registerModule(new AfterburnerModule())
+                .registerModule(new JavaTimeModule())
+                .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
```
Comment (schlosna) [inline extras/jackson-support/src/main/java/com/palantir/remoting1/ext/jackson/ObjectMappers.java:42]: do we have examples/tests of this?

File: extras/jackson-support/src/test/java/com/palantir/remoting1/ext/jackson/ObjectMappersTest.java (hunk 3 above)
Comment (uschi2000) [reply extras/jackson-support/src/test/java/com/palantir/remoting1/ext/jackson/ObjectMappersTest.java]: refactor typo, thanks

File: extras/jackson-support/src/main/java/com/palantir/remoting1/ext/jackson/ObjectMappers.java (hunk 4 above)
Comment (uschi2000) [reply extras/jackson-support/src/main/java/com/palantir/remoting1/ext/jackson/ObjectMappers.java:42]: yes. the ObjectMappersTests fail if you don't disable one of those features.

File: extras/jackson-support/src/test/java/com/palantir/remoting1/ext/jackson/ObjectMappersTest.java (hunk 2 above)
Comment (uschi2000) [reply extras/jackson-support/src/test/java/com/palantir/remoting1/ext/jackson/ObjectMappersTest.java]: we can either drop precision always or never.

Comment (uschi2000) [reply extras/jackson-support/src/test/java/com/palantir/remoting1/ext/jackson/ObjectMappersTest.java]: sorry, clarifying: either micro or nano precision.

File: extras/jackson-support/src/main/java/com/palantir/remoting1/ext/jackson/ObjectMappers.java (hunk 4 above)
Comment (schlosna) [reply extras/jackson-support/src/main/java/com/palantir/remoting1/ext/jackson/ObjectMappers.java:42]: Sorry, meant that we should probably have some tests for things like `java.time.LocalDate`

Comment (uschi2000) [reply extras/jackson-support/src/main/java/com/palantir/remoting1/ext/jackson/ObjectMappers.java:42]: Oh, I was hoping to get away with a token-test for OffsetDateTime and ZonedDateTime, but can add LocalDate too.

Comment (uschi2000) [reply extras/jackson-support/src/main/java/com/palantir/remoting1/ext/jackson/ObjectMappers.java:42]: Done

**Comment by uschi2000** [comment]
Anything left here?

**Pull request merged**